### PR TITLE
control/controlclient: remove quadratic allocs in mapSession

### PR DESCRIPTION
**Pull request by bradfitz**
The mapSession code was previously quadratic: N clients in a netmap
send updates proportional to N and then for each, we do N units of
work. This removes most of that "N units of work" per update. There's
still a netmap-sized slice allocation per update (that's #8963), but
that's it.

Bit more efficient now, especially with larger netmaps:

                                 │     before     │                after                │
                                 │     sec/op     │   sec/op     vs base                │
    MapSessionDelta/size_10-8       47.935µ ±  3%   1.232µ ± 2%  -97.43% (p=0.000 n=10)
    MapSessionDelta/size_100-8      79.950µ ±  3%   1.642µ ± 2%  -97.95% (p=0.000 n=10)
    MapSessionDelta/size_1000-8    355.747µ ± 10%   4.400µ ± 1%  -98.76% (p=0.000 n=10)
    MapSessionDelta/size_10000-8   3079.71µ ±  3%   27.89µ ± 3%  -99.09% (p=0.000 n=10)
    geomean                          254.6µ         3.969µ       -98.44%

                                 │     before     │                after                 │
                                 │      B/op      │     B/op      vs base                │
    MapSessionDelta/size_10-8        9.651Ki ± 0%   2.395Ki ± 0%  -75.19% (p=0.000 n=10)
    MapSessionDelta/size_100-8      83.097Ki ± 0%   3.192Ki ± 0%  -96.16% (p=0.000 n=10)
    MapSessionDelta/size_1000-8     800.25Ki ± 0%   10.32Ki ± 0%  -98.71% (p=0.000 n=10)
    MapSessionDelta/size_10000-8   7896.04Ki ± 0%   82.32Ki ± 0%  -98.96% (p=0.000 n=10)
    geomean                          266.8Ki        8.977Ki       -96.64%

                                 │    before     │               after                │
                                 │   allocs/op   │ allocs/op   vs base                │
    MapSessionDelta/size_10-8         72.00 ± 0%   20.00 ± 0%  -72.22% (p=0.000 n=10)
    MapSessionDelta/size_100-8       523.00 ± 0%   20.00 ± 0%  -96.18% (p=0.000 n=10)
    MapSessionDelta/size_1000-8     5024.00 ± 0%   20.00 ± 0%  -99.60% (p=0.000 n=10)
    MapSessionDelta/size_10000-8   50024.00 ± 0%   20.00 ± 0%  -99.96% (p=0.000 n=10)
    geomean                          1.754k        20.00       -98.86%

Updates #1909

### CODE REVIEW

File: control/controlclient/map.go [hunk 1]
```diff
@@ -198,32 +199,30 @@ func (ms *mapSession) HandleNonKeepAliveMapResponse(ctx context.Context, resp *t
...
+	changed int
 }
 
-// netmapForResponse returns a fully populated NetworkMap from a full
-// or incremental MapResponse within the session, filling in omitted
-// information from prior MapResponse values.
-func (ms *mapSession) netmapForResponse(resp *tailcfg.MapResponse) *netmap.NetworkMap {
-	undeltaPeers(resp, ms.previousPeers)
+// updateStateFromResponse updates ms from res. It takes ownership of res.
+func (ms *mapSession) updateStateFromResponse(res *tailcfg.MapResponse) {
```
Comment (maisem) [inline control/controlclient/map.go:213]: rename is a bit distracting

Comment (bradfitz) [reply control/controlclient/map.go:213]: Well, it's more than a rename. The func was broke in two halves, necessarily for a future change.

Comment (maisem) [reply control/controlclient/map.go:213]: i mean the `resp` -> `res` rename

Comment (bradfitz) [reply control/controlclient/map.go:213]: One half updates internal state only. (in the future some lite updates will only do that step, and not generate a big netmap).

The second half allocs the NetworkMap when needed. (currently always)

Comment (bradfitz) [reply control/controlclient/map.go:213]: Oh, I can put `resp` back.

File: control/controlclient/map.go [hunk 2]
```diff
@@ -63,25 +64,22 @@ type mapSession struct {
 
 	// Fields storing state over the course of multiple MapResponses.
 	lastNode               tailcfg.NodeView
+	peers                  map[tailcfg.NodeID]*tailcfg.NodeView // pointer to view (oddly). same pointers as sortedPeers.
+	sortedPeers            []*tailcfg.NodeView                  // same pointers as peers, but sorted.
```
Comment (maisem) [inline control/controlclient/map.go:68]: sorted by NodeID?

File: control/controlclient/map.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ type mapSession struct {
 	// Immutable fields.
 	nu             NetmapUpdater // called on changes (in addition to the optional hooks below)
 	privateNodeKey key.NodePrivate
```
Comment (maisem) [inline control/controlclient/map.go:35]: you're not setting this anymore

**Review by maisem — APPROVED** [review]
Nice!

**Pull request merged**